ui/plugin-system/src/components/TraceQueryEditor/TraceQueryEditor.tsx

```diff
@@ -0,0 +1,134 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { QueryDefinition } from '@perses-dev/core';
+import { Button, Stack } from '@mui/material';
+import AddIcon from 'mdi-material-ui/Plus';
+import { useState } from 'react';
+import { produce } from 'immer';
+import { usePlugin, TraceQueryDefinition } from '../../runtime';
+import { TraceQueryInput } from './TraceQueryInput';
+
+const DEFAULT_QUERY_PLUGIN_TYPE = 'TraceQuery';
+
+export interface TraceQueryEditorProps {
+  queries?: TraceQueryDefinition[];
+  onChange: (queries: QueryDefinition[]) => void;
+}
+
+export function TraceQueryEditor({ queries = [], onChange }: TraceQueryEditorProps) {
+  const hasMoreThanOneQuery = queries.length > 1;
+  const defaultTraceQueryKind = 'TempoTraceQuery';
+
+  // State for which queries are collapsed
+  // TODO: Would be easier if we had IDs for queries.
+  const [queriesCollapsed, setQueriesCollapsed] = useState(queries.map(() => false));
+
+  const { data: defaultQueryPlugin } = usePlugin(DEFAULT_QUERY_PLUGIN_TYPE, defaultTraceQueryKind, {
+    useErrorBoundary: true,
+    enabled: true,
+  });
+
+  // Query handlers
+  const handleQueryChange = (index: number, queryDef: TraceQueryDefinition) => {
+    onChange(
+      produce(queries, (draft) => {
+        if (draft) {
+          draft[index] = queryDef;
+        } else {
+          draft = [queryDef];
+        }
+      })
+    );
+  };
+
+  const handleQueryAdd = () => {
+    if (!defaultQueryPlugin) return;
+    onChange(
+      produce(queries, (draft) => {
+        const queryDef: TraceQueryDefinition = {
+          kind: DEFAULT_QUERY_PLUGIN_TYPE,
+          spec: {
+            plugin: { kind: defaultTraceQueryKind, spec: defaultQueryPlugin.createInitialOptions() },
+          },
+        };
+        if (draft) {
+          draft.push(queryDef);
+        } else {
+          draft = [...queries, queryDef];
+        }
+      })
+    );
+    setQueriesCollapsed((queriesCollapsed) => {
+      queriesCollapsed.push(false);
+      return [...queriesCollapsed];
+    });
+  };
+
+  const handleQueryDelete = (index: number) => {
+    onChange(
+      produce(queries, (draft) => {
+        draft.splice(index, 1);
+      })
+    );
+    setQueriesCollapsed((queriesCollapsed) => {
+      queriesCollapsed.splice(index, 1);
+      return [...queriesCollapsed];
+    });
+  };
+
+  const handleQueryCollapseExpand = (index: number) => {
+    setQueriesCollapsed((queriesCollapsed) => {
+      queriesCollapsed[index] = !queriesCollapsed[index];
+      return [...queriesCollapsed];
+    });
+  };
+
+  // show one query input if queries is empty
+  const queryDefinitions: TraceQueryDefinition[] = queries.length
+    ? queries
+    : [
+        {
+          kind: 'TraceQuery',
+          spec: {
+            plugin: {
+              kind: 'TempoTraceQuery',
```

Comment:
We might need a new package for the Tempo-specific parts of this that are in plugin-system, similar to how we have prometheus-plugin to encapsulate its query plugin, variables, datasource / client, editor, etc.

Reply:
@Nexucis since we don’t have a finished plugin system yet to install / manage third party plugins, should we just make a more generic package like `misc-plugins` for now? Or we could do a tempo-plugin but that probably won’t scale to add new packages like that in ui folder

Reply:
NVM sorry I missed that we have a tempo-plugin already now 😣